Cleaning up the python examples

nickkamp1 opened the pull request:
This PR is meant to clean up the six examples that will be presented in the paper, including:

- numu CC DIS in IceCube, DUNE and ATLAS
- dipole-portal HNLs in MiniBooNE, MINERvA, and CCM

The changes of this PR include:

- more sensible serialization of data output into both HDF5 and parquet files. Requires flattening the interaction records of the tree and including a single dataset for each property of the interaction record. This is done using awkward arrays
- convenience methods for requesting tabulated flux distribution on the python side, that interface with the literal flux data release for both the BNB and NuMI. Alex's HE SN flux is also supported in this way
- support for serialization of LIDarkNews{CrossSection,Decay} objects via pickle, including configuration methods that set up each object. The default implementation of DarkNews models in LI now loads these pickled objects rather than the numpy arrays of table data, though support for both methods still exists. 
- finalized paper plots for each example (notebooks plotting final state kinematic distributions and position distributions)

Things that were not addressed in this PR, but could/should be addressed at some point:
- clean up CCM example to sample from extended target source. should no longer support point source position distribution since it is incompatible with our assumed flux units. CCM is currently a special case (removed from scope of this PR)
- convenience methods for requesting CrossSection models, e.g. DIS or DarkNews (deemed not necessary)

Finally, methods have been added to initialize TabulatedFluxDistributions with vectors of doubles, though this is not currently used

austinschneider commented on .gitignore:
```diff
@@ -60,6 +63,18 @@
 *build
 
+# images
+*.png
+*/*.png
+*/*/*.png
+*/*/*/*.png
+*/*/*/*/*.png
+*.pdf
+*/*.pdf
+*/*/*.pdf
+*/*/*/*.pdf
+*/*/*/*/*.pdf
```
These lines can be reduced to just:
```.gitignore
*.png
*.pdf
```
since the .gitignore format will match these patterns in all folders.

austinschneider commented on .gitignore:
```diff
@@ -60,6 +63,18 @@
+*/*/*/*/*.pdf
+
 # pycache
 *pycache*
 */*pycache*
```
Similarly these lines can be reduced to just:
```.gitignore
__pycache__/
```
which will match only directories named `__pycache__`, and will do so in any folder

austinschneider commented on .gitignore:
```diff
@@ -60,6 +63,18 @@
 # build
 *build
```
This line should probably be changed to:
```.gitignore
/build/
```
which will only match a directory named `build` in the same folder as this `.gitignore` file

nickkamp1 replied:
yeah, this is because I was using a `cbuild` directory locally to test the manual cmake build

austinschneider commented on python/LIController.py:
```diff
@@ -215,12 +222,25 @@ def GetFiducialVolume(self):
-                if sector.name == fid_vol_dict[self.experiment]:
-                    fid_vol = sector.geo
-        return fid_vol
+        if self.experiment=="MiniBooNE":
+            return _geometry.Sphere(_geometry.Placement(_math.Vector3D([0,0,0])),
+                                    5.0, 0)
+        elif self.experiment=="CCM":
+            return _geometry.Cylinder(_geometry.Placement(_math.Vector3D([23,0,-0.65])),
+                                      0.96, 0, 1.232)
+        elif self.experiment=="MINERvA":
+            edges = [[0.000, 0.93675],
+                    [0.81125, 0.46838],
+                    [0.81125, -0.46838],
+                    [0.000, -0.93675],
+                    [-0.81125, -0.46838],
+                    [-0.81125, 0.46838]]
+            zsecs = [_geometry.ZSection(1.45,0,1),
+                     _geometry.ZSection(4.0,0,1)]
+            return _geometry.ExtrPoly(_geometry.Placement(_math.Vector3D([0,0,0])),
+                                      edges,zsecs)
+        else:
+            return None
```
I know we discussed where we wanted to put the fiducial volume information before, and decided it should probably be something user defined / in the code. But, if it's going to have some defaults that get used automatically via LIController, then it probably does make more sense to put them in the resources directory somewhere and version them. It would also make sense to provide the user with a mechanism to override these settings easily.

Maybe we could put them into the geometry .dat files as something that the DetectorModel does not parse, but that is retrieved by LIController? Or it could be in a separate folder since we already have folders for densitites and materials?

nickkamp1 replied:
This has been addressed in the one of the most recent commits—fiducial volumes are now defined in the geo .dat files

austinschneider commented on python/LIDarkNews.py:
```diff
@@ -374,43 +435,32 @@ def SaveInterpolationTables(self, total=True, diff=True):
-
     def GetPossiblePrimaries(self):
         return [Particle.ParticleType(self.ups_case.nu_projectile.pdgid)]
 
```
Dumb suggestion, but you could encapsulate this in a function.

```suggestion

    def _ensure_configured(self):
        if not self.is_configured:
            self.configure()
```

austinschneider commented on python/LIDarkNews.py:
```diff
@@ -374,43 +435,32 @@ def SaveInterpolationTables(self, total=True, diff=True):
         return [Particle.ParticleType(self.ups_case.nu_projectile.pdgid)]
 
     def GetPossibleTargetsFromPrimary(self, primary_type):
+        if not self.is_configured: self.configure()
```
```suggestion
        self._ensure_configured()
```

austinschneider commented on python/LIDarkNews.py:
```diff
@@ -374,43 +435,32 @@ def SaveInterpolationTables(self, total=True, diff=True):
 
     def GetPossibleTargets(self):
-        return [Particle.ParticleType(self.ups_case.nuclear_target.pdgid)]
+        if not self.is_configured: self.configure()
```
```suggestion
        self._ensure_configured()
```

austinschneider commented on python/LIDarkNews.py:
```diff
@@ -374,43 +435,32 @@ def SaveInterpolationTables(self, total=True, diff=True):
+        return [self.target_type]
 
     def GetPossibleSignatures(self):
+        if not self.is_configured: self.configure()
```
```suggestion
        self._ensure_configured()
```

austinschneider commented on python/__init__.py:
```diff
@@ -6,3 +6,22 @@
 from . import distributions
 from . import injection
+
+from . import _util
+
+# set up some public-facing utilities functions
+utilities.get_resource_package_dir = _util.resource_package_dir
+utilities.get_detector_model_path = _util.get_detector_model_path
+utilities.get_material_model_path = _util.get_material_model_path
+utilities.get_cross_section_model_path = _util.get_cross_section_model_path
+utilities.get_tabulated_flux_model_path = _util.get_tabulated_flux_model_path
+utilities.get_tabulated_flux_file = _util.get_tabulated_flux_file
+
+def darknews_version():
+    try:
+        import DarkNews
+        return _util.normalize_version(DarkNews.__version__)
+    except:
+        print("WARNING: DarkNews is not installed in the local environment")
+        return None
+utilities.darknews_version = darknews_version
```
Nice!

austinschneider commented on python/_util.py:
```diff
@@ -534,3 +535,21 @@ def get_material_model_path(model_name, must_exist=True):
+    return _get_model_path(model_name,prefix="Fluxes", is_file=False, must_exist=must_exist)
+ 
+ 
+def get_tabulated_flux_file(model_name, tag, must_exist=True):
+        abs_flux_dir = get_tabulated_flux_model_path(model_name,must_exist=must_exist)
+        # require existence of FluxCalculator.py
+        FluxCalculatorFile = os.path.join(abs_flux_dir,"FluxCalculator.py")
+        assert(os.path.isfile(FluxCalculatorFile))
+        spec = importlib.util.spec_from_file_location("FluxCalculator", FluxCalculatorFile)
+        FluxCalculator = importlib.util.module_from_spec(spec)
+        sys.modules["FluxCalculator"] = FluxCalculator
+        spec.loader.exec_module(FluxCalculator)
+        flux_file = FluxCalculator.MakeFluxFile(tag,abs_flux_dir)
+        del sys.modules["FluxCalculator"] # remove flux directory from the system
+        return flux_file
```
Very sneaky!

austinschneider commented on resources/Detectors/densities/DUNEFD/DUNEFD-v1.dat:
Good catch on this one!

austinschneider commented on resources/Examples/Example1/DIS_ATLAS.py:
```diff
@@ -0,0 +1,67 @@
+controller.Initialize()
+
+events = controller.GenerateEvents()
+
```
```suggestion

os.makedirs("output", exist_ok=True)

```

austinschneider commented on resources/Examples/Example1/DIS_DUNE.py:
```diff
@@ -0,0 +1,65 @@
+controller.Initialize()
+
+events = controller.GenerateEvents()
+
```
```suggestion

os.makedirs("output", exist_ok=True)

```

austinschneider commented on resources/Examples/Example1/DIS_IceCube.py:
```diff
@@ -67,4 +62,4 @@
 
 events = controller.GenerateEvents()
 
```
```suggestion

os.makedirs("output", exist_ok=True)

```

austinschneider commented on resources/Examples/Example2/DipolePortal_CCM.py:
```diff
@@ -88,9 +80,10 @@
 
-events = controller.GenerateEvents()
+events = controller.GenerateEvents(fill_tables_at_exit=False)
 
```
```suggestion

os.makedirs("output", exist_ok=True)

```

austinschneider commented on resources/Examples/Example2/DipolePortal_MINERvA.py:
```diff
@@ -88,9 +70,10 @@
 
-events = controller.GenerateEvents()
+events = controller.GenerateEvents(fill_tables_at_exit=False)
 
```
```suggestion

os.makedirs("output", exist_ok=True)

```

austinschneider commented on resources/Examples/Example2/DipolePortal_MiniBooNE.py:
```diff
@@ -88,9 +70,10 @@
 
-events = controller.GenerateEvents()
+events = controller.GenerateEvents(fill_tables_at_exit=False)
 
```
```suggestion

os.makedirs("output", exist_ok=True)

```

austinschneider commented on resources/Fluxes/NUMI/NUMI-v1.0/NUMI_RHC_LE.dat:
```diff
@@ -1,4 +1,4 @@
-#Enu(GeV) nue numu nutau nuebar numubar nutaubar (all fluxes in nus/cm^2/POT/GeV)
+Enu(GeV) nue numu nutau nuebar numubar nutaubar (all fluxes in nus/cm^2/POT/GeV)
```
Is removing the `#` comment character okay here?

nickkamp1 replied:
yup! we use the header information when grabbing the flux for each primary type

austinschneider reviewed (changes requested):
Looks great overall!

I suggested a few minor changes that should be easy to make and asked a question or two on some files.

I have not tested anything yet, but will try to do so shortly.

I'd say it's good to go with those small modifications and assuming the tests have been run!

nickkamp1 commented:
@austinschneider the most recent commits clean up a lot of the interpolation logic in DarkNews. Things also run a lot faster now. Can go over the details in our next meeting

nickkamp1 commented:
@austinschneider this PR is ready for review whenever you get the chance and then a rebase. I'm at the point now where I'm just working on the paper plots, so we should probably merge this PR

austinschneider reviewed (approved):
Looks good! Let me know when you're able to do the rebase or if you'd like me to do it instead.

nickkamp1 commented:
> Looks good! Let me know when you're able to do the rebase or if you'd like me to do it instead.

sounds good! let's do it in real time at our meeting

austinschneider reviewed (approved):
I think you were still going to add in some extra changes to fix that dark news issue, but otherwise it looks good.

nickkamp1 commented:
> I think you were still going to add in some extra changes to fix that dark news issue, but otherwise it looks good.

I added the changes to fix the dark news issue, and also added some more changes that address an issue in GetPossibleSignaturesForParents in the DarkNewsCrossSection class: the nucleon-level cross sections for DarkNews were not properly returning signatures when supplied with a nuclear target. This meant those interactions weren't being simulated properly, but that has now been fixed

austinschneider commented:
Still going through the DarkNews changes, but I think the changes to the way the fiducial volume is handled could cause issues down the line. I just drafted a PR into this branch that should fix things, but it probably needs some testing.
